bika/lims/adapters/referencewidgetvocabulary.py

```diff
@@ -164,6 +113,142 @@ def to_utf8(self, data):
                 self.to_utf8(key): self.to_utf8(value)
                 for key, value in data.iteritems()
             }
-            # if it's anything else, return it in its original form
 
+        # if it's anything else, return it in its original form
         return data
+
+    def get_query_from_request(self, name):
+        """Returns the query inferred from the request
+        """
+        query = self.request.get(name, "{}")
+        # json.loads does unicode conversion, which will fail in the catalog
+        # search for some cases. So we need to convert the strings to utf8
+        # https://github.com/senaite/senaite.core/issues/443
+        query = json.loads(query)
+        return self.to_utf8(query)
+
+    def get_raw_query(self):
+        """Returns the raw query to use for current search, based on the
+        base query + update query
+        """
+        query = self.base_query.copy()
+        search_query = self.search_query.copy()
+        query.update(search_query)
+
+        # Add sorting criteria
+        sorting = self.resolve_sorting(query)
+        query.update(sorting)
+
+        # Check if sort_on is an index and if is sortable. Otherwise, assume
+        # the sorting must be done manually
+        catalog = api.get_tool(self.catalog_name)
+        sort_on = query.get("sort_on", None)
+        if sort_on and not self.is_sortable_index(sort_on, catalog):
+            del(query["sort_on"])
+        return query
+
+    def resolve_sorting(self, query):
+        """Resolves the sorting criteria for the given query
+        """
+        sorting = {}
+
+        # Sort on
+        sort_on = query.get("sidx", None)
+        sort_on = sort_on or query.get("sort_on", None)
+        sort_on = sort_on == "Title" and "sortable_title" or sort_on
```

Comment:
Can we leave the "Title" instead of forcing "sortable_title"? The reason is that `sortable_title` is case sensitive which causes wierd sortings like `Z` items before `a` items

Reply:
Good point

Reply:
The problem is that "Title" is a `ZCTextIndex`, so not sortable:

```
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module bika.lims.browser.widgets.referencewidget, line 294, in __call__
  Module bika.lims.browser.widgets.referencewidget, line 264, in search
  Module bika.lims.adapters.referencewidgetvocabulary, line 243, in __call__
  Module bika.lims.adapters.referencewidgetvocabulary, line 198, in search
  Module Products.CMFPlone.CatalogTool, line 394, in searchResults
  Module bika.lims.monkey.zcatalog, line 64, in searchResults
  Module Products.ZCatalog.Catalog, line 916, in searchResults
  Module Products.ZCatalog.Catalog, line 894, in _getSortIndex
CatalogError: The index chosen for sort_on (Title) is not capable of being used as a sort index.
```

And one of the objectives of this refactoring was to rely on indexes and metadata as much as possible, and limit the results as well. If we have to do the sorting manually, then we have to search for all items first.. and for some portal types, this is going to be too expensive.

Reply:
Instead of doing workarounds (manual sorting, etc.) i think the best approach would be to make `sortable_title` index to store all values in lowercase. Is only a matter of adding a subscriber, although it means we'll have to do a reindex in upgrade step. Fortunately, sortable_title is only used in `setup_catalog`, so it shouldn't be a problem.

Reply:
Agree that it would be better to do the `sortable_title` index case insensitive. However, it is also used in the `analysis_catalog`: https://github.com/senaite/senaite.core/blob/master/bika/lims/catalog/analysis_catalog.py#L37
and `analysisrequest_catalog`: https://github.com/senaite/senaite.core/blob/master/bika/lims/catalog/analysisrequest_catalog.py#L34

Reply:
Worth to mention both Analysis and AnalysisCategory have indexers in place for `sortable_title` that are used to sort them correctly based on sortkey + title:

https://github.com/senaite/senaite.core/blob/master/bika/lims/catalog/indexers/baseanalysis.py#L26-L34

https://github.com/senaite/senaite.core/blob/master/bika/lims/catalog/indexers/analysiscategory.py#L25-L31

Reply:
So these indexers need to be changed:
- https://github.com/senaite/senaite.core/blob/master/bika/lims/catalog/indexers/baseanalysis.py#L27
- https://github.com/senaite/senaite.core/blob/master/bika/lims/catalog/indexers/analysiscategory.py#L26

Reply:
Done this case-insenstive dance for `sortable_title` here: https://github.com/senaite/senaite.core/pull/1337/commits/0481cd29cfc80814d37f40da58dc8cb1a79b279a